fdbserver/MoveKeys.actor.cpp

```diff
@@ -877,6 +881,60 @@ ACTOR Future<Void> removeStorageServer( Database cx, UID serverID, MoveKeysLock
 	}
 }
 
+ACTOR Future<Void> removeKeysFromFailedServer(Database cx, UID serverID, MoveKeysLock lock) {
+	state Key begin = allKeys.begin;
+	// Multi-transactional removal in case of large number of shards, concern in violating 5s transaction limit
+	while (begin < allKeys.end) {
+		state Transaction tr(cx);
+		loop {
+			try {
+				tr.info.taskID = TaskPriority::MoveKeys;
+				tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+				wait(checkMoveKeysLock(&tr, lock));
+				TraceEvent("RemoveKeysFromFailedServerLocked")
+				    .detail("ServerID", serverID)
+				    .detail("Version", tr.getReadVersion().get())
+				    .detail("Begin", begin);
+				// Get all values of keyServers and remove serverID from every occurrence
+				// Very inefficient going over every entry in keyServers
+				// No shortcut because keyServers and serverKeys are not guaranteed same shard boundaries
+				state KeyRange currentKeys = KeyRangeRef(begin, allKeys.end);
+				state Standalone<RangeResultRef> keyServers =
+				    wait(krmGetRanges(&tr, keyServersPrefix, currentKeys,
+				                      SERVER_KNOBS->MOVE_KEYS_KRM_LIMIT, SERVER_KNOBS->MOVE_KEYS_KRM_LIMIT_BYTES));
+				for (auto it : keyServers) {
+					vector<UID> src;
+					vector<UID> dest;
+					decodeKeyServersValue(it.value, src, dest);
+
+					// The failed server is not present
+					if (std::find(src.begin(), src.end(), serverID) == src.end() &&
+					    std::find(dest.begin(), dest.end(), serverID) == dest.end()) {
+						continue;
+					}
+
+					// Update the vectors to remove failed server then set the value again
+					// Dest is usually empty, but keep this in case there is parallel data movement
+					src.erase(std::remove(src.begin(), src.end(), serverID), src.end());
+					dest.erase(std::remove(dest.begin(), dest.end(), serverID), dest.end());
+					tr.set(keyServersKey(it.key), keyServersValue(src, dest));
+				}
+
+				// Set entire range for our serverID in serverKeys keyspace to false to signal erasure
+				wait(krmSetRangeCoalescing(&tr, serverKeysPrefixFor(serverID), currentKeys, allKeys, serverKeysFalse));
```

Comment:
This will violate the invariant: the server must be removed from key range [a, k] in `keyServers` list before the server's responsible shards in [a,k] can be removed from `serverKeys` list.

In other words,  in the multi-transactional situation, the current implementation will erasure "all keys"  for the serverID in `serverKeys` list before "all keys" are removed from `keyServers` list. 

This may be the reason why `DeliveredToNotAssigned ` error may happen.